File: projects/biomemetics/labs/mantis/scripts/environment/py/mantis_environment/schema_check.py
"""Python Draft 2020-12 checker for the shared environment schema fixtures."""

from __future__ import annotations

import json
from pathlib import Path
import sys

from jsonschema import Draft202012Validator, FormatChecker


def load(path: Path) -> object:
    return json.loads(path.read_text(encoding="utf-8"))


def main(argv: list[str] | None = None) -> int:
    args = argv if argv is not None else sys.argv[1:]
    if len(args) != 3:
        print("usage: schema_check.py SCHEMA POSITIVE NEGATIVE", file=sys.stderr)
        return 64
    schema_path, positive_path, negative_path = map(Path, args)
    schema = load(schema_path)
    Draft202012Validator.check_schema(schema)
    validator = Draft202012Validator(schema, format_checker=FormatChecker())
    positive_errors = list(validator.iter_errors(load(positive_path)))
    negative_errors = list(validator.iter_errors(load(negative_path)))
    if positive_errors:
        print("positive fixture must validate", file=sys.stderr)
        for error in positive_errors:
            print(error.message, file=sys.stderr)
        return 1
    if not negative_errors:
        print("negative fixture must not validate", file=sys.stderr)
        return 1
    print("python jsonschema: positive ok, negative rejected")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: projects/biomemetics/labs/mantis/scripts/environment/ts/schema-check.mjs
#!/usr/bin/env node
// Zero-dependency checker for the shared environment JSON Schema fixtures.
// Applies the locked rules from schema.json to the positive and negative instances.

import { readFileSync } from "node:fs";

function fail(message) {
  process.stderr.write(`${message}\n`);
  process.exit(1);
}

if (process.argv.length !== 5) {
  fail("usage: schema-check.mjs SCHEMA POSITIVE NEGATIVE");
}

const [, , schemaPath, positivePath, negativePath] = process.argv;
const schemaText = readFileSync(schemaPath, "utf8");
if (
  !schemaText.includes("environment-fixture") ||
  !schemaText.includes('"additionalProperties": false') ||
  !schemaText.includes('"minimum": 0')
) {
  fail("schema fixture does not declare the locked environment-fixture rules");
}

const schema = JSON.parse(schemaText);
if (schema.type !== "object" || schema.additionalProperties !== false) {
  fail("schema type/additionalProperties mismatch");
}

function validate(instance) {
  if (instance === null || typeof instance !== "object" || Array.isArray(instance)) {
    throw new Error("instance must be an object");
  }
  const keys = Object.keys(instance);
  for (const key of keys) {
    if (key !== "kind" && key !== "value") {
      throw new Error(`additional property: ${key}`);
    }
  }
  if (instance.kind !== "environment-fixture") {
    throw new Error("kind must be the const environment-fixture");
  }
  if (!Number.isInteger(instance.value) || instance.value < 0) {
    throw new Error("value must be an integer >= 0");
  }
}

try {
  validate(JSON.parse(readFileSync(positivePath, "utf8")));
} catch (error) {
  fail(`positive fixture must validate: ${error.message}`);
}

let negativeRejected = false;
try {
  validate(JSON.parse(readFileSync(negativePath, "utf8")));
} catch {
  negativeRejected = true;
}
if (!negativeRejected) {
  fail("negative fixture must not validate");
}

process.stdout.write("typescript schema-check: positive ok, negative rejected\n");
